File: GitToolBox/src/zielu/gittoolbox/cache/RepoInfo.java
package zielu.gittoolbox.cache;

import com.google.common.base.Objects;
import com.google.common.base.Optional;
import zielu.gittoolbox.status.GitAheadBehindCount;

public class RepoInfo {
    public final RepoStatus status;
    public final Optional<GitAheadBehindCount> count;

    private RepoInfo(RepoStatus _status, Optional<GitAheadBehindCount> _count) {
        status = _status;
        count = _count;
    }

    public static RepoInfo create(RepoStatus status, Optional<GitAheadBehindCount> count) {
        return new RepoInfo(status, count);
    }

    @Override
    public String toString() {
        return Objects.toStringHelper(this)
            .add("status", status)
            .add("count", count)
            .toString();
    }
}

File: src/main/kotlin/zielu/gittoolbox/cache/RepoInfo.kt
package zielu.gittoolbox.cache

import zielu.gittoolbox.status.GitAheadBehindCount
import java.util.Optional

internal data class RepoInfo(
  val status: RepoStatus,
  val count: GitAheadBehindCount?,
  val tags: List<String>
) {
  fun isEmpty(): Boolean = status.isEmpty && count == null

  fun tagsNotEmpty(): Boolean = tags.isNotEmpty()

  @Deprecated("for Java compatibility", replaceWith = ReplaceWith("count?.") )
  fun maybeCount(): Optional<GitAheadBehindCount> = Optional.ofNullable(count)

  companion object {
    private val empty: RepoInfo = RepoInfo(RepoStatus.empty(), null, listOf())

    @JvmStatic
    fun create(
      status: RepoStatus,
      count: GitAheadBehindCount?,
      tags: Collection<String>
    ): RepoInfo {
      return RepoInfo(status, count, tags.toSet().toList())
    }

    @JvmStatic
    fun empty(): RepoInfo = empty
  }
}
